src/main/java/org/commcare/core/network/ModernHttpRequester.java

```diff
@@ -72,26 +82,35 @@ private void setupAuthentication(boolean isAuth, User user, String domain) {
                 username = user.getUsername();
             }
             final String password = user.getCachedPwd();
-            if (username == null || password == null || User.TYPE_DEMO.equals(user.getUserType())) {
-                String message =
-                        "Trying to make authenticated http request without proper credentials";
-                throw new RuntimeException(message);
-            } else if (!"https".equals(url.getProtocol())) {
-                throw new PlainTextPasswordException();
-            } else {
-                Authenticator.setDefault(new Authenticator() {
-                    @Override
-                    protected PasswordAuthentication getPasswordAuthentication() {
-                        return new PasswordAuthentication(username, password.toCharArray());
-                    }
-                });
-            }
+            setupAuthentication(username, password, user);
         } else {
             // clear any prior set authenticator to make unauthed requests
             Authenticator.setDefault(null);
         }
     }
 
+    private void setupAuthentication(final String username, final String password, User user) {
+        if (username == null || password == null ||
+                (user != null && User.TYPE_DEMO.equals(user.getUserType()))) {
```

Comment:
So if you look at the full changes to this class, that check for demo user was already there. The `ModernHttpRequester` class is used for a lot more than just this feature, it's just a generic HTTP requester. All that I've done here is to add a way to use this class to make a request from a user who hasn't already logged into CommCare, by creating this overloaded method that accepts just `username` and `password` String arguments, instead of a `User` object (which is only accessible once a user has logged in).

At the level of this class, there's also no distinction between a mobile user and a web user, we're just thinking in terms of authentication for an HTTP request, so we ultimately just need a username and password. If you read through the old implementation of `setupAuthentication()` that method accepted a `User` object, and all that that object was being used for was to grab the username and password from it, and then it used that to set up the `Authenticator`. So I've moved all of that work to my overloaded method, and changed the original version to simply call my new version with the username and password taken from the `User` object.

If this is still confusing, we can talk through in person, which might be easier.